docs/database_management.md

```diff
@@ -0,0 +1,18 @@
+### Database Management
+To run a local version of this project you will need to start a 
+repl and initialize the vanilla_db.
+This can be done in 
+[vanilla.db.core](src/clj/vanilla/db/core.clj)
+on line 132 (As of 02/10/2020), and run the following line of code from the repl  to initialize a local
+copy of the database.
+```
+(initialize-database vanilla-db)
+```
+
+When this project is deployed, it packages up vanilla_default (the database 
```

Comment:
dson't we need to run the initializer against vanilla-default in order to create the default that can be copied over during the dockerfile/deploy? I think we need a bit more clarity on "For Local Development" use vanilla-db and "For Updateing the Repo" use vanilla-default...

Reply:
Okay,  I'm a little confused about the whole database thing.
First let me say I haven't dug into how we're doing our db stuff, nor into the SQLite universe.
Partly I'm reaching back into my knowledge of how we did things with Oracle or SQLServer.

So firstly there needs to be some script, DDL,whatever, that can generate an EMPTY db with the structure we want in the project.

Second, I don't think there should be an 'initialize db' available in production code - unless its behind a comment() so we can play in the repl.

Third, there needs to be an empty db available to use for continuing development, OR the 'initialize db' script needs to be documented well enough to create the empty db.

Fourth, there needs to be a 'known good' db in the repo for production and for development that doesn't involve db updates.

So based on that, I would think the 'vanilla default' db would be the empty db, although I'd rather just document using initialize db (or rather 'create db') to make that.

The 'vanilla db' should be the agreed upon starting point for the project based on successfully merged PRs.  That will get copied and used in production code.

Is this what we're moving towards, or am I completely off-base?

Reply:
You are absolutely correct. 

1. to run the server without crashing we need a db with all the tables created and any initial data the server requires, such as the services table. the server needs this to be called _vanilla-db_
2. (initialize-database ...) can create such a db
3. since this db gets loaded with user data as we use the system, we need a "pristine" db in the repo so we can build the docker container and deploy. this will be called _vanilla-default_
4. to add new services to Vanilla, new rows must be added to the services table. to do this, we modify (initialize-database ...) so the data gets built correctly. then you should run (initialize-database ...) in the repl to generate the "new" _vanilla-default_ (footnote 1)
5. to test locally, you need to copy vanilla-default to vanilla-db so the server app can find it (or you can just use the "alternate" version of (initialize-database ...) at the repl and make _vanilla-db_ directly)
6. as you test, vanilla-db will become cluttered with your test data (logins, layouts, etc) so we do NOT want to have it in the repo, hence putting it into .gitignore

The trick now is to get this to be explained in the Docs and the code, so developers know how to do the right thing.

(footnote 1) in the both Chocolate and the sample apps, Luminus actually created "migration" files for use by a lein plugin to create the "default db" but we don't (yet) have that for Vanilla

Reply:
Chris answered all of those questions the way I would have.

To reiterate, the "known good" database is the vanilla_default. It never gets written to or read from during the projects runtime. It is just a clean database that docker and deployment use.

I will add more clarifying comments to the above documentation based on this comment chain.

Also, would it be worse to have the system create this database during lein run vs a lein plugin that we make? I was adding code yesterday that checks if vanilla_db exist, and if it doesn't create it from an (initialize-db) call; otherwise, just continue through standard execution.